### PR TITLE
use std::vector instead of prefvec

### PR DESCRIPTION
**Pull request by RvDijk**

### CODE REVIEW

**Comment by RvDijk** [comment]
@blsemo thanks for letting me know, I didn't know that before. I'll try to rewrite this PR s.t. it will have the std::vector simply as a member.

**Comment by blsemo** [comment]
That would be very cool, thanks!

On 28 October 2017 09:21:18 BST, Ruben van Dijk <notifications@github.com> wrote:
>@blsemo thanks for letting me know, I didn't know that before. I'll try
>to rewrite this PR s.t. it will have the std::vector simply as a
>member.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/SpaceGroupUCL/depthmapX/pull/82#issuecomment-340159187

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by blsemo** [comment]
Thanks, that is already better. The question now is whether a PointMaps class needs to exist at all (I don't think so) and whether Metagraph should be derived from PointMaps (most certainly not) - the main problem is that the architecture of the existing code base is very weird, and we are slowly working on fixing that.
If you look at it, pointmaps has
- a vector of pointmaps
- an index which is the currently displayed point map 
- a super space pix (the function of which isn't quite clear to me at the moment, need to investigate...)
- some methods to add/handle pointmaps

The only place that class is used is as super class to MetaGraph - which is a code smell in my opinion...
I think the cleanest solution would be if Metagraph just had a std::vector of PointMap and stored which one is the current one, and the methods of PointMaps can either be deleted if they aren't used, or can become just functions that are called from metagraph.
What we should also add are unit tests that verify that those functions do what it says on the tin...

PS:
If you wonder why fixing the architecture matters - currently, it is very hard to reason about what the code actually does, and it is especially hard to reason about the memory layout (because everything is a container that derives from another class that either has some containers or derives from some other container). And I have a very strong suspicion that one of the performance pain points of depthmap is bad cache locality - i.e. if it is running its analysis in a tight loop, the data it needs to read and write is spread all over the heap, and it has to access the RAM a lot rather than having everything in the processor cache. This can easily slow things down by an order of magnitude or two...

PPS:
I'll also comment on some spots in the code where more modern C++ idioms can/should be used.

File: depthmapX/GraphDoc.cpp [hunk 1]
```diff
@@ -938,7 +938,7 @@ void QGraphDoc::OnEditGrid()
    }
    bool newmap = false;
 
-   if (!m_meta_graph->PointMaps::size() || m_meta_graph->getDisplayedPointMap().isProcessed()) {
+   if (!m_meta_graph->PointMaps::maps_vector.size() || m_meta_graph->getDisplayedPointMap().isProcessed()) {
```
Comment (blsemo) [inline depthmapX/GraphDoc.cpp]: That fact that you have to write PointMaps::maps_vector in a file talking to the MetaGraph class is a fairly strong code smell - I know it's been there before, but that just means the old code wasn't good, and we want to improve it when we touch it...

File: salalib/pointdata.cpp [hunk 2]
```diff
@@ -131,18 +131,18 @@ int PointMaps::addNewMap(const std::string& name)
    bool duplicate = true;
    while (duplicate) {
       duplicate = false;
-      for (size_t i = 0; i < size(); i++) {
-         if (at(i).getName() == myname) {
+      for (size_t i = 0; i < maps_vector.size(); i++) {
```
Comment (blsemo) [inline salalib/pointdata.cpp]: for loops on standard containers can be written more concise and safer as
`for (auto& current : maps_vector)`

**Comment by RvDijk** [comment]
@blsemo that sounds reasonable. I think that, for me, implementing all of this is going to be a bit out of the scope of my job description. With that being said, I don't mind looking at this for some more time.  However, when it is going to take too much time, then someone else should take over this PR. 
I could look into including the std::vector in Metagraph and removing the pointmaps class. 
For the unit tests, I think that writing these require actual understanding of the depthmapX program, which I really haven't. On top of that, I don't have experience in writing unit tests (yet).

**Pull request closed without merging**